### ISSUE TITLE
plugins.youtube Refer to #3724 "Unable to open URL: https://youtube.com/get_video_info" when using "--retry-streams"

### ISSUE THREAD

**Issue by xanek** [streamlink/streamlink#3779]
## Plugin Issue

- [X] This is a plugin issue and I have read the contribution guidelines.
- [X] I am using the latest development version from the master branch.

### Description

Continuation of the issue of #3724 , problem sprung up again despite it being temporarily fixed for a few weeks.

### Reproduction steps / Explicit stream URLs to test

Refer to #3724 
Any URL for a YouTube livestream that has not gone live yet.

### Log output

```
C:\>streamlink --loglevel debug
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.8.7
[cli][debug] Streamlink: 2.1.2+24.g5579278
[cli][debug] Requests(2.25.1), Socks(1.7.1), Websocket(1.0.1)
[cli][debug] Arguments:
[cli][debug]  --loglevel=debug
[cli][debug]  --rtmpdump=C:\Program Files (x86)\Streamlink\rtmpdump\rtmpdump.exe
[cli][debug]  --ffmpeg-ffmpeg=C:\Program Files (x86)\Streamlink\ffmpeg\ffmpeg.exe
```
### Additional comments

Same errors as before
```
[cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=wQtZrQj8AHQ&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=wQtZrQj8AHQ&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=wQtZrQj8AHQ&html5=1&el=detailpage)
```

**Edit:** 
Note, this happens on upcoming streams that have not gone live yet, not streams that have already been archived or deleted.

For example, on the upcoming WWDC apple live stream that's in 2 days in June 7 (https://www.youtube.com/watch?v=0TD96VTf0Xs) , I get the error, whether it be within 10 seconds of trying or 2+ minutes.

![image](https://user-images.githubusercontent.com/23298441/120909188-1e139480-c638-11eb-98fe-1c295ed6ccbb.png)

Please note this is something that I have been doing for months, with "--retry-streams" without problem until the first occurrence, which led to #3724 issue thread to be made and then a minor fix that worked for 2 weeks before the opening of this issue thread.

The command I am using for streamlink for streams to record is
`streamlink --retry-streams 2 -o <filename>.ts <youtube link> <quality>`
so for example with the Apple WWDC, it would be 
`streamlink --retry-streams 2 -o test.ts https://www.youtube.com/watch?v=0TD96VTf0Xs best`

What is supposed to happen is that after retrying x amount of times in y amount of seconds, it is supposed to just eventually start recording and outputting the file when the stream does become live when it checks via the --retry-streams and that should be all it should be doing.

But instead after a certain period it will give the errors and ignore recording even after a stream goes live until the command is interrupted and tried again.

**Comment by back-to** [comment]
@xanek 

could you try https://gist.githubusercontent.com/back-to/182ad20960fedd5ead8691deea064a6e/raw/67b58f7ac58386475565bb8d5c42f7aa9d3b5e73/youtube.py and report back if it helps

**Comment by xanek** [comment]
> @xanek
> 
> could you try https://gist.githubusercontent.com/back-to/182ad20960fedd5ead8691deea064a6e/raw/67b58f7ac58386475565bb8d5c42f7aa9d3b5e73/youtube.py and report back if it helps

Same errors unfortunately.

**Comment by back-to** [comment]
to fix this, you will need to use a **cookie** from your browser

get the cookie `VISITOR_INFO1_LIVE` and use it with `--http-cookie VISITOR_INFO1_LIVE=XXXXXXX`

https://streamlink.github.io/cli.html#cmdoption-http-cookie

**Comment by eremem** [comment]
Providing the cookie's value from the browser doesn't seem to change anything:

`$ /usr/bin/streamlink --http-cookie VISITOR_INFO1_LIVE=here-goes-my-cookie https://www.youtube.com/watch?v=XDD78TUI2ZE best`
`[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=XDD78TUI2ZE
error: Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=XDD78TUI2ZE&el=detailpage)`

**Comment by xanek** [comment]
> to fix this, you will need to use a **cookie** from your browser
> 
> get the cookie `VISITOR_INFO1_LIVE` and use it with `--http-cookie VISITOR_INFO1_LIVE=XXXXXXX`
> 
> https://streamlink.github.io/cli.html#cmdoption-http-cookie

Okay, I have yet to test this and you closed the issue?

**Comment by xanek** [comment]
> to fix this, you will need to use a **cookie** from your browser
> 
> get the cookie `VISITOR_INFO1_LIVE` and use it with `--http-cookie VISITOR_INFO1_LIVE=XXXXXXX`
> 
> https://streamlink.github.io/cli.html#cmdoption-http-cookie

Okay, I have yet to test this and you closed the issue without even confirming with me or others that it works though?
I am currently testing it for 30 minutes to see if it shows this issue.
But @eremem reply seems to show that it doesn't work at the moment?

**Comment by xanek** [comment]
> > to fix this, you will need to use a **cookie** from your browser
> > get the cookie `VISITOR_INFO1_LIVE` and use it with `--http-cookie VISITOR_INFO1_LIVE=XXXXXXX`
> > https://streamlink.github.io/cli.html#cmdoption-http-cookie
> 
> Okay, I have yet to test this and you closed the issue without even confirming with me or others that it works though?
> I am currently testing it for 30+ minutes to see if it shows this issue.
> But @eremem reply seems to show that it doesn't work at the moment?

@back-to 
This does nothing, please re-open the issue.
![image](https://user-images.githubusercontent.com/23298441/121792396-d487f400-cbb9-11eb-9a5f-bdf46523372e.png)

**Comment by xanek** [comment]
> > to fix this, you will need to use a **cookie** from your browser
> > get the cookie `VISITOR_INFO1_LIVE` and use it with `--http-cookie VISITOR_INFO1_LIVE=XXXXXXX`
> > https://streamlink.github.io/cli.html#cmdoption-http-cookie
> 
> Okay, I have yet to test this and you closed the issue without even confirming with me or others that it works though?
> I am currently testing it for 30+ minutes to see if it shows this issue.
> But @eremem reply seems to show that it doesn't work at the moment?

@back-to 
This does nothing, I tested it by letting it run for 5 minutes and it happened again, please re-open the issue.
![image](https://user-images.githubusercontent.com/23298441/121792401-e4073d00-cbb9-11eb-9f7b-c3e1b132f58a.png)

**Comment by xanek** [comment]
> to fix this, you will need to use a **cookie** from your browser
> 
> get the cookie `VISITOR_INFO1_LIVE` and use it with `--http-cookie VISITOR_INFO1_LIVE=XXXXXXX`
> 
> https://streamlink.github.io/cli.html#cmdoption-http-cookie

@back-to 
Can you reopen the issue? Your solution didn't work, did you even test your solution?
![image](https://user-images.githubusercontent.com/23298441/121792474-c4244900-cbba-11eb-8913-ef8eb3693fcc.png)

```
G:\>streamlink --http-cookie VISITOR_INFO1_LIVE=cookies.txt --retry-streams 2 -o test.ts https://www.youtube.com/watch?v=UlYxSlx0EfM best
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=UlYxSlx0EfM
[cli][info] Waiting for streams, retrying every 2.0 second(s)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=embedded)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=detailpage)
Interrupted! Exiting...
```

**Comment by xanek** [comment]
> to fix this, you will need to use a **cookie** from your browser
> 
> get the cookie `VISITOR_INFO1_LIVE` and use it with `--http-cookie VISITOR_INFO1_LIVE=XXXXXXX`
> 
> https://streamlink.github.io/cli.html#cmdoption-http-cookie

@back-to
Can you reopen the issue? Your solution didn't work, did you even test your solution? This happens after minute+ of testing it with cookies.

![image](https://user-images.githubusercontent.com/23298441/121792586-a35cf300-cbbc-11eb-84e5-3b50ceb5a303.png)

Unless there is something with the cookies I need to specifically do? Unless I am using the cookies incorrectly that I grabbed using an extension from chrome?
Can you explain how to use this command with the cookie? There is no specifics listed on the site and I can only assume to use a textfile of cookies used that is taken from a browser extension that gets cookies.

**Comment by mkbloke** [comment]
@xanek, please stop adding the same comments over and over again.

The `--http-cookie` option does not take a cookies.txt file.  It's used as: `--http-cookie cookie_name=cookie_value`

**Comment by bastimeyer** [comment]
You are using the `--http-cookie` argument incorrectly. Check the debug log. Both `--http-cookie` and `--http-header` arguments accept key-value pairs in the format `key=value`. They don't read data from files like `--config` does.
https://streamlink.github.io/cli.html#cmdoption-http-cookie

Apart from that, the plugin is working fine without custom cookies when retrying offline streams, unless you've got blocked by YT, which will happen when (ab)using their services for a while with tools like streamlink.

```
$ streamlink -l debug --stdout --retry-streams 2 'https://www.youtube.com/watch?v=UlYxSlx0EfM' best >/dev/null
[cli][debug] OS:         Linux-5.12.10-1-git-x86_64-with-glibc2.33
[cli][debug] Python:     3.9.5
[cli][debug] Streamlink: 2.1.2+35.g6cdc3eb
[cli][debug] Requests(2.25.1), Socks(1.7.1), Websocket(0.58.0)
[cli][debug] Arguments:
[cli][debug]  url=https://www.youtube.com/watch?v=UlYxSlx0EfM
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][debug]  --stdout=True
[cli][debug]  --retry-streams=2.0
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=UlYxSlx0EfM
[plugins.youtube][debug] Video ID from URL
[plugins.youtube][debug] Using video ID: UlYxSlx0EfM
[plugins.youtube][debug] get_video_info - 1: Diese Live-Veranstaltung beginnt in 11 Stunden.
[plugins.youtube][debug] get_video_info - 2: Diese Live-Veranstaltung beginnt in 11 Stunden.
[plugins.youtube][debug] get_video_info - 3: Diese Live-Veranstaltung beginnt in 11 Stunden.
[cli][info] Waiting for streams, retrying every 2.0 second(s)
[plugins.youtube][debug] Video ID from URL
[plugins.youtube][debug] Using video ID: UlYxSlx0EfM
[plugins.youtube][debug] get_video_info - 1: Diese Live-Veranstaltung beginnt in 11 Stunden.
[plugins.youtube][debug] get_video_info - 2: Diese Live-Veranstaltung beginnt in 11 Stunden.
[plugins.youtube][debug] get_video_info - 3: Diese Live-Veranstaltung beginnt in 11 Stunden.
[plugins.youtube][debug] Video ID from URL
[plugins.youtube][debug] Using video ID: UlYxSlx0EfM
[plugins.youtube][debug] get_video_info - 1: Diese Live-Veranstaltung beginnt in 11 Stunden.
[plugins.youtube][debug] get_video_info - 2: Diese Live-Veranstaltung beginnt in 11 Stunden.
[plugins.youtube][debug] get_video_info - 3: Diese Live-Veranstaltung beginnt in 11 Stunden.
^CInterrupted! Exiting...
```

Since I've not been blocked by YT, I can't validate that the cookie @back-to has mentioned works, but he probably won't have posted it without having it tested himself.

**Comment by xanek** [comment]
How do I get the `key=value` then to test if it works for me then?

**Comment by mkbloke** [comment]
Depends on the browser.  In FF/Chrome: Key F12 > Storage > Cookies.  Find desired cookie in list and click on it.  Copy name and then value from top of right hand pane to get key/value pair.

**Comment by bastimeyer** [comment]
https://github.com/streamlink/streamlink/issues/3779#issuecomment-860099900

Open the link in your web browser, open the dev-tools (F12), go the network tab, press ctrl+f, search for the cookie name, find any of the resulting HTTP requests and copy the cookie value from the HTTP headers field, which is in the format `cookie: key1=value1; key2=value2; ...`.

**Comment by xanek** [comment]
I am unsure which portion I am looking at to get the cookie value
![image](https://user-images.githubusercontent.com/23298441/121792956-73afea00-cbc0-11eb-9661-76d9aa75d4b9.png)

**Comment by mkbloke** [comment]
Search for `VISITOR_INFO1_LIVE` is what bastimeyer meant.  Or see my revised comment for another method.

**Comment by xanek** [comment]
So is it just the very short value shown in this image?

![image](https://user-images.githubusercontent.com/23298441/121793146-cc33b700-cbc1-11eb-948d-40def68ac478.png)

**Comment by mkbloke** [comment]
Yes, that's the one.

**Comment by xanek** [comment]
It doesn't seem to work for me
`G:\>streamlink --http-cookie KEY=[my value, which was 11 characters long] --retry-streams 2 -o test.ts https://youtu.be/UlYxSlx0EfM best
[cli][info] Found matching plugin youtube for URL https://youtu.be/UlYxSlx0EfM
[cli][info] Waiting for streams, retrying every 2.0 second(s)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&eurl=https%3A%2F%2Fyoutube.googleapis.com%2Fv%2FUlYxSlx0EfM)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://www.youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=UlYxSlx0EfM&html5=1&el=detailpage)`

**Comment by bastimeyer** [comment]
I'm going to lock this now, as it's getting a bit out of hand. This is a bug tracker and not a help tracker for issues that are out of scope of this project. If you don't know what HTTP cookies are and how they work, then please google.

One final explicit example (without explicit value) for the suggested solution. If it doesn't work, then there's nothing that can be done here.

```
streamlink --loglevel debug --http-cookie 'VISITOR_INFO1_LIVE=YOUR_VALUE_HERE' 'https://youtu.be/UlYxSlx0EfM' best
```
